src/instruction.rs

```diff
@@ -1296,6 +1302,52 @@ impl Instruction {
             nom::combinator::all_consuming(parse_instruction)(&lexed).map_err(|e| e.to_string())?;
         Ok(instruction)
     }
+
+    /// Per the language spec, whether this instruction's timing within the pulse
+    /// program must be precisely controlled, so as to begin exactly on the end of
+    /// the latest preceding timed instruction
+    pub(crate) fn requires_precise_timing(&self) -> bool {
+        match self {
+            Instruction::Capture(_)
+            | Instruction::Delay(_)
+            | Instruction::Fence(_)
+            | Instruction::Pulse(_)
+            | Instruction::RawCapture(_) => true,
+            Instruction::Arithmetic(_)
+            | Instruction::BinaryLogic(_)
+            | Instruction::CalibrationDefinition(_)
+            | Instruction::CircuitDefinition(_)
+            | Instruction::Convert(_)
+            | Instruction::Comparison(_)
+            | Instruction::Declaration(_)
+            | Instruction::Exchange(_)
+            | Instruction::FrameDefinition(_)
+            | Instruction::Gate(_)
+            | Instruction::GateDefinition(_)
+            | Instruction::Halt
```

Comment:
A query on the spec; wouldn't a HALT have a precise timing? ie. you would know precisely when the program ends?

Reply:
You make a good point, but it may have more to do with the docs or choice of names.

Many instructions will have timing fixed at compile time, but whether a halt is fixed to time X or time Y does not materially affect the program, if we see its pulse sequence as essential to its purpose.

Put another way, depending on the semantics of a hardware provider, the HALT may be 10us or 10ms after the previous instruction, but that does not affect the readout results that one would expect to collect.

That’s in contrast to a PULSE, where durations and times may be statically (if parametrically) determined, and which do not depend on hardware provider implementation details.

How could I better communicate that, maybe in docs but ideally in the code itself?

Reply:
I think the documentation. As a user, I never write HALT; it's always implicit, and if the hardware back end chooses to add it for some tracking of the type you describe, so be it. The place I would care is Quil-T based pulse timing analysis. One of the things we can do with Quil-T entirely on the user side is mimic pulse timing rules and determine the timing of all of the instructions ourselves. Here, we report total program duration as the end time of the last pulse. But, where HALT is in play (implicitly or explicitly) and has some real time duration as you describe, that real time should be accounted for (start time of HALT was the end time of the last pulse, but then there is some real end time to HALT that is greater?). Which brings me back to the point-- wouldn't you be able to tell me precisely what that time was? If the back end adds 10 microseconds, shouldn't I be able to see that somehow? Is this was "precise" means here, and what is `quil-rs`'s part to play in this workflow (it may be it is simply out of scope at this level).

This was a query of the semantics only, and not a practical issue, so I'm happy whatever you decide.

Reply:
Thanks for that feedback, @mhodson-rigetti !

Following up from our discussion offline, I've renamed this method to `is_scheduled`, and `ExecutionDependency::Timing` to `ExecutionDependency::Scheduled`.

For anyone following along - "precise timing" vs scheduling is a worthwhile semantic distinction here; only some instructions are considered for "scheduling" as part of the program, and their position within that schedule is essential to the program doing what the author intended. That is now what this function identifies.

All instructions, though, may have "precise times" fixed as part of compilation for a backend. The timing of non-scheduled instructions (like `ADD` or `MOVE`) is flexible within some bounds without affecting the output of the program.